src/main/kotlin/no/nav/modiapersonoversikt/storage/S3StorageProvider.kt

```diff
@@ -25,29 +32,82 @@ class S3StorageProvider : StorageProvider {
                 .withCredentials(AWSStaticCredentialsProvider(credentials)).build()
-        createBucketIfMissing()
+
+        lagS3BucketsHvisNodvendig(NOTIFIKASJON_BUCKET_NAME, SIST_LEST_BUCKET_NAME)
+    }
+
+    override fun hentNotifikasjoner(ident: String): List<NotifikasjonDTOOut> {
+        return timed("hent_notifikasjoner") {
+            val sistLest = hentSistLest(ident)
+            val meldinger: List<NotifikasjonDTOOut> = hentAlleNotifikasjoner()
+
+            meldinger.map { it.copy(settTidligere = it.opprettetDato.isBefore(sistLest)) }
+        }
     }
 
-    override fun getMeldinger(): List<Melding> {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    override fun hentNotifikasjon(id: UUID): NotifikasjonDTOOut? {
+        return timed("hent_notifikasjon") {
+            val meldingContent = s3.getObject(NOTIFIKASJON_BUCKET_NAME, id.toString())
+            val lagretMelding = gson.fromJson(InputStreamReader(meldingContent.objectContent), Notifikasjon::class.java)
+            NotifikasjonDTOOut(id, lagretMelding)
+        }
+    }
+
+    override fun opprettNotifikasjon(notifikasjon: NotifikasjonDTOIn): UUID {
+        return timed("opprett_notifikasjon") {
+            val id = UUID.randomUUID()
+            val notifikasjonTilLagring = Notifikasjon(notifikasjon, LocalDateTime.now())
+            s3.putObject(NOTIFIKASJON_BUCKET_NAME, id.toString(), gson.toJson(notifikasjonTilLagring))
+            id
+        }
+    }
+
+    override fun slettNotifikasjon(id: UUID) {
+        timed("slett_notifikasjon") {
+            s3.deleteObject(NOTIFIKASJON_BUCKET_NAME, id.toString())
+        }
     }
 
-    override fun getMelding(id: UUID): Melding {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    override fun oppdaterSistLest(ident: String) {
+        timed("oppdater_sist_lest") {
+            s3.putObject(SIST_LEST_BUCKET_NAME, ident, DATE_FORMATTER.format(LocalDateTime.now()))
+        }
     }
 
-    override fun putMelding(melding: Melding): UUID {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    private fun hentSistLest(ident: String): LocalDateTime {
+        return timed("hent_sist_lest") {
+            try {
+                val sistLest = s3.getObjectAsString(SIST_LEST_BUCKET_NAME, ident)
+
+                LocalDateTime.parse(sistLest, DATE_FORMATTER)
+            } catch (e: Exception) {
+                MIN_DATE
+            }
+        }
     }
 
-    override fun fjernMelding(id: UUID) {
-        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    private fun hentAlleNotifikasjoner(): List<NotifikasjonDTOOut> {
+        return timed("hent_alle_notifikasjoner") {
+            try {
+                s3.listObjectsV2(NOTIFIKASJON_BUCKET_NAME)
```

Comment:
Om noen vet om enn bedre møte å hente ut alle objekter på så skrik ut. :P 
Nå gjøres det N + 1 kall til S3 for å hente innholdet fra alle N notifikasjoner.

Totalt sett så vil det derfor skje N + 2 kall mot S3 for hver side last. :/

Reply:
Vi gjør det så enkelt som mulig først og så får vi heller optimalisere seinere.